README.md
# cnode-search-cli

Command Line for search topics on Cnodejs.org from baidu

[![NPM](https://nodei.co/npm/cnode-search-cli.png?downloads=true&stars=true)](https://nodei.co/npm/cnode-search-cli/)
## Installation
```
npm install -g cnode-search-cli
```

## Usage
```
Usage: cnode-search <keyword>

Options:

  -h, --help     output usage information
  -V, --version  output the version number
```
## License
MIT


CHANGELOG.md
0.0.11 / 2015-7-10
* 操作时添加颜色
* 退出从`exit`改为 `q`
* 操作多次

0.0.1 / 2015-6-16
------------------
* Initial release.


lib/url.js
'use strict';

/**
 * 百度搜索url抽象
 * @param keyword 关键词
 * @param domain 域名
 * @constructor
 */
function Baidu(keyword, domain) {
  if (keyword) {
    this.keyword = encodeURI(keyword);
  }

  //搜索前缀
  this.searchPrefix = 'https://www.baidu.com/s?wd=';

  if (domain) {
    //需要限定搜索的域名范围
    this.domain = (typeof domain !== 'undefined' && domain !== '') ? ['site%3A', domain, '%20'].join('') : '';
  }

  //跳过条数，10条为一页
  this.pn = 0;

  this.pageSize = 10;
}

/**
 * 设置关键字
 * @param keyword
 */
Baidu.prototype.setKeyword = function(keyword) {
  if (keyword) {
    this.keyword = encodeURI(keyword);
  }
};

/**
 * 获得当前第几页
 * @returns {number}
 */
Baidu.prototype.currentPage = function() {
  return (this.pn / this.pageSize) + 1;
};

/**
 * 下一页
 * @returns {number}
 */
Baidu.prototype.nextPage = function() {
  return (this.pn / this.pageSize) + 2;
};

/**
 * 上一页
 * @returns {number}
 */
Baidu.prototype.prePage = function() {
  return (this.pn / this.pageSize) > 0 ?
    (this.pn / this.pageSize) + 1:
    1;
};

/**
 * 当前页url
 * @returns {string}
 */
Baidu.prototype.currentPageUrl = function() {
  return [
    this.searchPrefix,
    this.domain,
    this.keyword,
    '&pn=',
    (this.currentPage() - 1) * this.pageSize
  ].join('');
};

/**
 * 下一页地址
 * @returns {string}
 */
Baidu.prototype.nextPageUrl = function() {
  return [
    this.searchPrefix,
    this.domain,
    this.keyword,
    '&pn=',
    this.currentPage() * this.pageSize
  ].join('');
};

/**
 * 上一页地址
 * @returns {string}
 */
Baidu.prototype.prePageUrl = function() {
  return [
    this.searchPrefix,
    this.domain,
    this.keyword,
    '&pn=',
    this.currentPage() === 1 ? 0 : (this.currentPage() - 2) * this.pageSize
  ].join('');
};

exports.Baidu = Baidu;


lib/search.js
'use strict';

var openurl = require("openurl");
var request = require("request");
var cheerio = require('cheerio');
var Spinner = require('cli-spinner').Spinner;
var colors = require('colors/safe');
var Baidu = require('./url').Baidu;
var stdin = process.stdin;
var stdout = process.stdout;

stdin.setEncoding('utf8');
stdin.on('data', handler);

var domain = 'cnodejs.org';
var options = {
  url: '',
  headers: {
    'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.152 Safari/537.36'
  }
};
var baidu = new Baidu(null, domain);
//存放查询结果列表
var pages = [];


/**
 * 搜索并解析html
 * @param {String} keyword 搜索关键字
 * @param {Function} cb 回调
 */
function getDocumentUrl(keyword, cb) {
  baidu.setKeyword(keyword);
  options.url = baidu.currentPageUrl();

  request(options, function(err, res, body) {
    if (err) return cb(err);

    if (body == '') {
      return cb(colors.red('搜索失败=>: ' + keyword));
    }

    var $ = cheerio.load(body, {normalizeWhitespace: true});
    var $titleArr = $('.result > h3.t > a');
    console.log('\n=> ' + colors.green(keyword) + '\n');
    $titleArr.each(function(i, item){
      var title = $(item).text();
      var link = item.attribs.href;
      var reg = new RegExp(keyword, 'ig');
      console.log(['  ', colors.green((i + 1)), '  ', title.replace(reg, colors.green(keyword))].join(''));
      pages.push({title: title, link: link});
    });

    cb(null);
  });
}

/**
 * 取得最终页面地址并在浏览器中打开
 * 由于百度加入了302跳转，需要拿到最终页面地址
 * @param {Object} options url和userAgent
 */
function getTargetUrl(options) {
  request(options, function (err, res, body) {
    var target = ['http://', res.request.host, res.request.path].join('');
    console.log(target);
    openurl.open(target, function(err) {
      if (err) {
        console.log('there is no browser or platform is not support.');
        process.exit(0);
      }
    });
    ask();
  });
}

/**
 * ask for operation
 */
function ask() {
  console.log('');
  var question = [
    '操作:\n',
    '  1.选择一个序号并打开\n',
    '  2.输入\'q\'退出\n'
  ];
  stdout.write(colors.gray(question.join('')));
  stdin.resume();
}

/**
 * 命令行输入时的handler
 * @param data
 */
function handler(data) {
  if (isNaN(data)) {
    //if input 'q' then exit
    if (data.replace(/^\s+|\s+$/g, '').indexOf('q') === -1) {
      console.log(colors.red('暂时不兹瓷此操作'));
      ask();
    } else {
      process.exit(0);
    }
  } else if ( data < 1 || data > (pages.length + 1)) {
    console.log(colors.red('序号请在1-10中选择'));
  } else {
    stdin.pause();
    console.log('=> ' + pages[data - 1].title);
    options.url = pages[data - 1].link;
    getTargetUrl(options);
  }
}

module.exports = function search(keyword) {
  var spinner = new Spinner('%s');
  spinner.setSpinnerString('|/-\\');
  spinner.start();

  getDocumentUrl(keyword, function(err) {
    spinner.stop('clean');

    if (err) {
      console.log(err);
      process.exit(1);
    }

    ask();
  });
};

lib/cnode-search.js
'use strict';

var openurl = require("openurl");
var request = require("request");
var cheerio = require('cheerio');
var Spinner = require('cli-spinner').Spinner;
var colors = require('colors/safe');
var stdin = process.stdin;
var stdout = process.stdout;

stdin.setEncoding('utf8');
stdin.on('data', handler);

var options = {
  url: '',
  headers: {
    'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.152 Safari/537.36'
  }
};
//store results form baidu
var pages = [];
//maybe the number of pages to skip in baidu's pagination
//first page is 0,pn of second page is 10
var pn = 0;

/**
 * search and parse the result html
 * @param {String} keyword keyword to search
 * @param {Function} cb callback function
 */
function getDocumentUrl(keyword, cb) {
  var searchPrefix = 'https://www.baidu.com/s?wd=site%3Acnodejs.org%20';
  options.url = searchPrefix + encodeURI(keyword) + '&pn=' + pn;

  request(options, function(err, res, body) {
    if (err) return cb(err);

    if (body == '') {
      return cb(colors.red('Can not search keyword from cnode: ' + keyword));
    }

    var $ = cheerio.load(body, {normalizeWhitespace: true});
    var $titleArr = $('.result > h3.t > a');
    console.log(colors.cyan('\nResults of search by: ') + colors.green(keyword) + '\n');
    $titleArr.each(function(i, item){
      var title = $(item).text();
      var link = item.attribs.href;
      console.log(['  ', colors.cyan((i + 1)), '  ', title].join(''));
      pages.push({title: title, link: link});
    });

    cb(null);
  });
}

/**
 * get target page link and open it on browser
 * @param {Object} options url and Agent
 */
function getTargetUrl(options) {
  request(options, function (err, res, body) {
    var target = 'http://' + res.request.host + res.request.path;
    console.log(target);
    openurl.open(target);
    ask();
  });
}

/**
 * ask for operation
 */
function ask() {
  console.log('');
  var question = [
    'Operation:\n',
    '  1.Choice a number of page to open\n',
    '  2.\'q\' to quit.\n'
  ];
  stdout.write(colors.yellow(question.join('')));
  stdin.resume();
}

/**
 * operation handler
 * @param data
 */
function handler(data) {
  if (isNaN(data)) {
    //if input 'q' then exit
    if (data.replace(/^\s+|\s+$/g, '').indexOf('q') === -1) {
      console.log(colors.red('This operation is not support.'));
      ask();
    } else {
      process.exit(0);
    }
  } else if ( data < 1 || data > (pages.length + 1)) {
    console.log(colors.red('Page number must between 1 and ' + ( pages.length + 1 ) + '.'));
  } else {
    stdin.pause();
    console.log('=> ' + pages[data - 1].title);
    options.url = pages[data - 1].link;
    getTargetUrl(options);
  }
}

module.exports = function search(keyword) {
  var spinner = new Spinner('%s');
  spinner.setSpinnerString('|/-\\');
  spinner.start();

  getDocumentUrl(keyword, function(err) {
    spinner.stop('clean');

    if (err) {
      console.log(err);
      process.exit(1);
    }

    ask();
  });
};